Fix changes to InputValidator to resolve accidental breaking changes

Nokel81 opened the pull request:
Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix changes to InputValidator to resolve accidental breaking changes

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Show changes working by removing no longer required empty props objects

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix type errors

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jweak reviewed (dismissed):
LGTM

Nokel81 committed:
Fix async validation errors not being displayed

- And fix validation errors sometimes being displayed multiple times

Signed-off-by: Sebastian Malton <sebastian@malton.name>

aleksfront commented on src/renderer/components/+extensions/install.tsx:
```diff
@@ -30,17 +28,22 @@ interface Dependencies {
-  InputValidators.isUrl,
-  InputValidators.isPath,
-  InputValidators.isExtensionNameInstall,
-];
```
It looks more readable to keep all validators in array comparing to new version where different logic mixed in one function.

Nokel81 replied:
The problem is that the old code was had a bug in it. It would cause a "UnhandledPromiseRejectionError" because `isPath` is an async validator

aleksfront replied:
Well, I still think this new function `installInputValidator` should be simplified or splitted to smaller ones. If we want to add some new validator to the list, we will touch code of this `installInputValidator` and probably add some another ifs or replace conditions which is not good. Not good because it can break other things.

Maybe we can use `for ... of` to iterate through the validators array https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of

My quick checks shows that it "handles" async and non-async items:

```
const x = () => true
const y = () => true
const z = () => {
    return new Promise(resolve => {
        resolve(true);
    });
}

for await (const variable of [x, y, z]) {
    console.log(variable())
}

// prints
// VM593:2 true
// VM593:2 Promise {<fulfilled>: true}
```

Nokel81 replied:
The problem is that in the async case, it isn't that the promise should resolve to `true` but that it shouldn't reject. This is a hold over from how they were previously implemented.

ixrock replied:
Haven't read everything in this PR.. but just quick question:
if we don't want to that some async validator _might throw exception_ why we cannot just wrap it with extra `.catch(() => true)` or something?

Nokel81 replied:
The thing is that async validators have always used `throw` to mean "failed to validate". This PR does not change that and doesn't try to. Also, the error message from the rejection was always used as the message displayed to the user.

I agree that a better and more unified design could have been fixed (either always using throw/reject or always using `return boolean`).

ixrock replied:
If I'm not mistaken if we wrap every promise with extra `.catch(() => {})` we can filter all relevant errors / skip   `UnhandledPromiseRejectionError` or just passthrough inside catch with `throw` again.

e.g. something like this wouldn't work?

```ts
// validation cycle
await Promise.resolve(validator()).catch(err => {
   if(err instanceof UnhandledPromiseRejectionError) return true; // all good
   throw String(err); // pass real validation error
})
```

Nokel81 replied:
No unfortunately because we won't get back `UnhandledPromiseRejectionError`.

What I think I can do though, is make a "union" function for joining several input validators together.

ixrock replied:
What about catching this error with `process.on('unhandledRejection')`? Would it make sense?

```ts
process.on('unhandledRejection', error => {
  // Will print "unhandledRejection err is not defined"
  console.log('unhandledRejection', error.message);
});

new Promise((_, reject) => reject(new Error('woops'))).
  catch(error => {
    // Will not execute
    console.log('caught', err.message);
  });
```

Nokel81 replied:
That seems very hacky and wrong, especially since that is global shared state. It isn't that hard to make a union function that works correctly. Just shows the horribleness of the current design.

Nokel81 replied:
I have pushed the unionizing function, PTAL.

aleksfront commented on src/renderer/components/input/input.tsx:
```diff
@@ -168,32 +175,16 @@ export class Input extends React.Component<InputProps, State> {
-            error => this.getValidatorError(value, validator) || error,
-          ),
-        );
-      }
```
Quite heavy code block removed, which is super great. But to make 100% sure we don't need this anywhere?

Nokel81 replied:
Is caused the duplicate error messages bug. Let me find the issue about that.

Nokel81 replied:
I cannot find it, but here is a screenshot of the bug (from master)
![Screen Shot 2022-05-17 at 3 47 33 PM](https://user-images.githubusercontent.com/8225332/168897537-ee45aead-845d-4542-a6d3-bf0529c1f7a1.png)

aleksfront commented on src/renderer/components/input/input_validators.ts:
```diff
@@ -92,26 +116,24 @@ export const isExtensionNameInstall = inputValidator({
 });
 
-export const minLength = inputValidator({
+export const minLength = inputValidator<true>({
```
First time see this notation, does it mean same as `inputValidator<boolean>` or it can't be `false`?

Nokel81 replied:
It means that the second parameter for the validate function is required. I can make that easier to read.

Nokel81 replied:
I have backed out the ability to specify that it is required, thus making this easier to read.

aleksfront reviewed (commented):
Can we specify what exact issues it solves and where is the actual fixes / additional improvements? I'm just a bit worried that we're adding extra code here.

Nokel81 committed:
Simplify builders for validators

- Replace the boolean type params on `function inputValidator` with a
  new builder `function inputValidatorWithRequiredProps` to make the
  code easier to read

Signed-off-by: Sebastian Malton <sebastian@malton.name>

chenhunghan commented on src/renderer/components/+extensions/install.tsx:
```diff
@@ -30,17 +28,22 @@ interface Dependencies {
+      InputValidators.isUrl.validate(value)
+      || InputValidators.isExtensionNameInstall.validate(value)
+    ) {
+      return;
```
Do we need to return `true`? A bit confusing for me if a validator doesn't return boolean..

Nokel81 replied:
Fair enough, the rule is that synchronous validators must return booleans but async validators reject to show errors.

chenhunghan commented on src/renderer/components/input/input.tsx:
```diff
@@ -153,7 +161,7 @@ export class Input extends React.Component<InputProps, State> {
     const value = this.getValue();
     let validationId = (this.validationId = ""); // reset every time for async validators
-    const asyncValidators: Promise<any>[] = [];
+    const asyncValidators: Promise<React.ReactNode>[] = [];
```
Could you please explain why is validators an array of `Promise<React.ReactNode>`? I went through the code but don't get why they are `Promise<React.ReactNode>`

Nokel81 replied:
The reason is on line 182 below. Basically this is here to signify that the promises added to this list should resolve to something that can be rendered (the error message)

chenhunghan commented on src/renderer/components/input/input_validators.ts:
```diff
@@ -4,37 +4,59 @@
+
+export function inputValidator(validator: InputValidator<false, false>): InputValidator<false, false> {
+  return validator;
+}
```
I don't get how these two different.. just the name? They seems to get the same input, and return the same output.

chenhunghan replied:
ah, now I found the diff, `InputValidator<false, false>` vs `InputValidator<true, false>`

chenhunghan commented on src/renderer/components/input/input_validators.ts:
```diff
@@ -92,30 +114,26 @@ export const isExtensionNameInstall = inputValidator({
-      await fse.pathExists(value);
-    } catch {
-      throw new AsyncInputValidationError(`${value} is not a valid file path`);
+    if (!await fse.pathExists(value)) {
```
Looks like old code catches the error thrown from `fse.pathExists`, do we still need to catch this exception, or it's handled somewhere else?

Nokel81 replied:
The old code was wrong `fse.pathExists` never throws, it always resolves.

Nokel81 committed:
Introduce unionizing functions for input validators

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
fix tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

panuhorsmalahti commented on src/renderer/components/input/input_validators.ts:
```diff
@@ -4,41 +4,125 @@
+       * If no validator returns `true` then mark as invalid by throwing. The message will be
+       * obtained from the `message` field.
+       */
+      throw {};
```
Why not `throw new Error();`?

Nokel81 replied:
Sure, it doesn't matter.

aleksfront reviewed (commented):
I still got the errors trying to compile extension which uses validators. Compilation goes fine with previous `k8slens/extensions` version `5.5.0-git.c376b848c9.0`. Error example:
![validator errors](https://user-images.githubusercontent.com/9607060/172315578-01098ed0-e5fd-4d39-977c-5fced280eca8.png)

I guess we have to decide between 2 options:
1. Fine-tune this PR to remove extension compilation errors (sine the name of the PR contains "resolve accidental breaking changes").
2. Just to change typings within our extension to match them provided in current PR.

Nokel81 committed:
Remove RequiredProps type param

Signed-off-by: Sebastian Malton <sebastian@malton.name>

chenhunghan reviewed (approved):
LGTM(?)
